File: labs/mobile/donor-archive/tunnel/WireguardIpamManager.kt
package com.luminet.android.tunnel

data class WireguardPeer(
    val publicKey: String,
    val assignedIp: String,
    val allowedIps: List<String>,
    var endpoint: String? = null,
    var enabled: Boolean = true
)

class WireguardIpamManager(val subnetPrefix: String, val listenPort: Int) {
    private var nextSuffix: Int = 2
    val serverIp: String = "$subnetPrefix.1"
    private val peers = HashMap<String, WireguardPeer>()
    private val ipToKey = HashMap<String, String>()

    init {
        val parts = subnetPrefix.split(".")
        if (parts.size != 3) {
            throw IllegalArgumentException("Subnet prefix must have 3 octets, e.g. 10.14.0")
        }
    }

    fun allocatePeer(publicKey: String): WireguardPeer {
        peers[publicKey]?.let { return it }

        if (nextSuffix >= 254) {
            throw IllegalStateException("Subnet exhausted")
        }

        val assignedIp = "$subnetPrefix.$nextSuffix"
        nextSuffix++

        val peer = WireguardPeer(
            publicKey = publicKey,
            assignedIp = assignedIp,
            allowedIps = listOf("$assignedIp/32")
        )

        peers[publicKey] = peer
        ipToKey[assignedIp] = publicKey
        return peer
    }

    fun releasePeer(publicKey: String) {
        val peer = peers.remove(publicKey) ?: throw NoSuchElementException("Peer not found")
        ipToKey.remove(peer.assignedIp)
    }

    fun generateServerConfig(serverPrivateKey: String): String {
        val lines = ArrayList<String>()
        lines.add("[Interface]")
        lines.add("Address = $serverIp/24")
        lines.add("ListenPort = $listenPort")
        lines.add("PrivateKey = $serverPrivateKey")
        lines.add("")

        for (peer in peers.values) {
            if (peer.enabled) {
                lines.add("[Peer]")
                lines.add("PublicKey = ${peer.publicKey}")
                lines.add("AllowedIPs = ${peer.allowedIps.joinToString(", ")}")
                peer.endpoint?.let { lines.add("Endpoint = $it") }
                lines.add("")
            }
        }

        return lines.joinToString("\n")
    }

    fun generateClientConfig(
        peerPublicKey: String,
        peerPrivateKey: String,
        serverPublicKey: String,
        serverEndpoint: String
    ): String {
        val peer = peers[peerPublicKey] ?: throw NoSuchElementException("Peer not registered")

        val lines = ArrayList<String>()
        lines.add("[Interface]")
        lines.add("Address = ${peer.assignedIp}/32")
        lines.add("PrivateKey = $peerPrivateKey")
        lines.add("DNS = 1.1.1.1")
        lines.add("")

        lines.add("[Peer]")
        lines.add("PublicKey = $serverPublicKey")
        lines.add("Endpoint = $serverEndpoint")
        lines.add("AllowedIPs = 0.0.0.0/0, ::/0")
        lines.add("PersistentKeepalive = 25")

        return lines.joinToString("\n")
    }
}

File: labs/control-ui/api-shadow/wireguard_ipam_manager.ts
export interface WireguardPeer {
  publicKey: string;
  assignedIp: string;
  allowedIps: string[];
  endpoint?: string;
  enabled: boolean;
}

export class WireguardIpamManager {
  private subnetPrefix: string;
  private nextIpSuffix: number = 2;
  public serverIp: string;
  public listenPort: number;
  private peers: Map<string, WireguardPeer> = new Map();
  private ipToKey: Map<string, string> = new Map();

  constructor(subnetPrefix: string, listenPort: number) {
    const parts = subnetPrefix.split('.');
    if (parts.length !== 3) {
      throw new Error('Subnet prefix must have 3 octets, e.g. 10.14.0');
    }
    this.subnetPrefix = subnetPrefix;
    this.serverIp = `${subnetPrefix}.1`;
    this.listenPort = listenPort;
  }

  allocatePeer(publicKey: string): WireguardPeer {
    const existing = this.peers.get(publicKey);
    if (existing) {
      return { ...existing };
    }

    if (this.nextIpSuffix >= 254) {
      throw new Error('Subnet exhausted');
    }

    const assignedIp = `${this.subnetPrefix}.${this.nextIpSuffix}`;
    this.nextIpSuffix += 1;

    const peer: WireguardPeer = {
      publicKey,
      assignedIp,
      allowedIps: [`${assignedIp}/32`],
      enabled: true,
    };

    this.ipToKey.set(assignedIp, publicKey);
    this.peers.set(publicKey, peer);
    return { ...peer };
  }

  releasePeer(publicKey: string): void {
    const peer = this.peers.get(publicKey);
    if (!peer) {
      throw new Error('Peer not found');
    }
    this.ipToKey.delete(peer.assignedIp);
    this.peers.delete(publicKey);
  }

  getPeer(publicKey: string): WireguardPeer | undefined {
    const p = this.peers.get(publicKey);
    return p ? { ...p } : undefined;
  }

  generateServerConfig(serverPrivateKey: string): string {
    const lines: string[] = [];
    lines.push('[Interface]');
    lines.push(`Address = ${this.serverIp}/24`);
    lines.push(`ListenPort = ${this.listenPort}`);
    lines.push(`PrivateKey = ${serverPrivateKey}`);
    lines.push('');

    for (const peer of this.peers.values()) {
      if (peer.enabled) {
        lines.push('[Peer]');
        lines.push(`PublicKey = ${peer.publicKey}`);
        lines.push(`AllowedIPs = ${peer.allowedIps.join(', ')}`);
        if (peer.endpoint) {
          lines.push(`Endpoint = ${peer.endpoint}`);
        }
        lines.push('');
      }
    }

    return lines.join('\n');
  }

  generateClientConfig(
    peerPublicKey: string,
    peerPrivateKey: string,
    serverPublicKey: string,
    serverEndpoint: string
  ): string {
    const peer = this.peers.get(peerPublicKey);
    if (!peer) {
      throw new Error('Peer not registered');
    }

    const lines: string[] = [];
    lines.push('[Interface]');
    lines.push(`Address = ${peer.assignedIp}/32`);
    lines.push(`PrivateKey = ${peerPrivateKey}`);
    lines.push('DNS = 1.1.1.1');
    lines.push('');

    lines.push('[Peer]');
    lines.push(`PublicKey = ${serverPublicKey}`);
    lines.push(`Endpoint = ${serverEndpoint}`);
    lines.push('AllowedIPs = 0.0.0.0/0, ::/0');
    lines.push('PersistentKeepalive = 25');

    return lines.join('\n');
  }
}
